reference-tests/src/main/java/net/consensys/linea/MapFailedReferenceTestsTool.java

```diff
@@ -0,0 +1,170 @@
+
+import java.io.FileWriter;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import lombok.Synchronized;
+import lombok.extern.slf4j.Slf4j;
+import net.consensys.linea.zktracer.json.JsonConverter;
+
+@Slf4j
+public class MapFailedReferenceTestsTool {
+
+  public static void mapAndStoreFailedReferenceTest(
+      String testName, List<String> logEventMessages, String jsonOutputPath) {
+    Set<String> failedConstraints = getFailedConstraints(logEventMessages);
+    String jsonString = readFailedTestsOutput(jsonOutputPath);
+    JsonConverter jsonConverter = JsonConverter.builder().build();
+
+    List<ModuleToConstraints> constraintToFailingTests =
+        getModuleToConstraints(jsonString, jsonConverter);
+
+    mapFailedConstraintsToTestsToModule(failedConstraints, constraintToFailingTests, testName);
+
+    jsonString = jsonConverter.toJson(constraintToFailingTests);
+    writeToJsonFile(jsonString, jsonOutputPath);
+  }
+
+  private static void mapFailedConstraintsToTestsToModule(
+      Set<String> failedConstraints,
+      List<ModuleToConstraints> constraintToFailingTests,
+      String testName) {
+    for (String constraint : failedConstraints) {
+      String moduleName = getModuleFromFailedConstraint(constraint);
+      if (constraintToFailingTests.stream()
+          .filter(mapping -> mapping.equals(moduleName))
+          .toList()
+          .isEmpty()) {
+        constraintToFailingTests.add(new ModuleToConstraints(moduleName, new HashMap<>()));
+      }
+
+      ModuleToConstraints moduleMapping =
+          constraintToFailingTests.stream()
+              .filter(mapping -> mapping.equals(moduleName))
+              .toList()
+              .getFirst();
+
+      String cleanedConstraintName = constraint;
+      if (cleanedConstraintName.contains(".")) {
+        cleanedConstraintName = cleanedConstraintName.split("\\.")[1];
+      }
+      Set<String> failedTests =
+          new HashSet<>(
+              moduleMapping
+                  .constraints()
+                  .getOrDefault(cleanedConstraintName, Collections.emptyList()));
+      failedTests.add(testName);
+      moduleMapping.constraints().put(cleanedConstraintName, failedTests.stream().toList());
+    }
+  }
+
+  public static List<ModuleToConstraints> getModuleToConstraints(
+      String jsonString, JsonConverter jsonConverter) {
+    List<ModuleToConstraints> constraintToFailingTests = new ArrayList<>();
+    if (!jsonString.isEmpty()) {
+      constraintToFailingTests =
+          new ArrayList<>(
+              Arrays.asList(jsonConverter.fromJson(jsonString, ModuleToConstraints[].class)));
+    }
+    return constraintToFailingTests;
```

Comment:
the name of the method seems inconsistent with the returned value. Wouldn't it be cleared `getConstraintsToFailingTests`?